File: 2021/day-6.py
import copy

all_fish = [3,4,3,1,2]

file = open('input-6.txt', 'r')
file_lines = file.readlines()
all_fish = [ int(value) for value in file_lines[0].split(',')]

# print(all_fish)

# for day in range(1,19):
# 	for index in range(0, len(all_fish)):
# 		fish = all_fish[index]
# 		if fish == 0:
# 			all_fish[index] = 6
# 			all_fish.append(8)
# 		else:
# 			all_fish[index] -= 1
	# print("Day: {}".format(day))
	# print(all_fish)

# print(len(all_fish))

# For part 2, we won't maintain a list of fish but a list of counters for 0 to 8 that will increment as time goes
counters = [0, 0, 0, 0, 0, 0, 0, 0, 0] # index 0 to 8
temp_counters = copy.deepcopy(counters)
# instantiate based on input
for fish in all_fish:
	counters[fish] += 1

# print(counters)

# loop through the days
for day in range(1,257):
	for index, counter in enumerate(counters):
		if index == 0:
			temp_counters[6] = counter
			temp_counters[8] = counter
		elif index == 7:
			temp_counters[6] += counter # there are 2 ways you might reach 6, either from 0 or from 7 - need to account for this
		else:
			temp_counters[index-1] = counter 
	counters = copy.deepcopy(temp_counters)

sum = 0
for counter in counters:
	sum += counter


print(sum)

File: 2022/day-6.rb
data = File.open("input-6").read

index = 0
length = 4
while index+length < data.length
  four_chars = data[index..index+length-1]
  if four_chars.count(four_chars[0]) == 1 && four_chars.count(four_chars[1]) == 1 && four_chars.count(four_chars[2]) == 1 && four_chars.count(four_chars[3]) == 1
    break
  else
    index += 1
  end
end

puts index + length   

# part 2, we'll need to refine this shit and use detect

index = 0
length = 14
while index+length < data.length
  four_chars = data[index..index+length-1].split('')
  if four_chars.detect{ |e| four_chars.count(e) > 1 }
    index += 1
  else
    break
  end
end

puts index + length   